app/javascript/components/ansible-playbook-edit-catalog-form/index.jsx

```diff
@@ -0,0 +1,217 @@
+        });
+      });
+  }, [initialData.catalogFormId]);
+
+  const onCancel = () => {
+    const msg = sprintf(__('Edit of Catalog Item %s was cancelled by the user'), data.formData.description);
+    // eslint-disable-next-line no-undef
+    miqFlashLater({ message: msg });
+    window.location.href = '/catalog/explorer?';
+  };
+
+  const onSubmit = (values) => {
+    const requestObject = prepareRequestObject(values);
+    const redirectUrl = `/catalog/explorer/${initialData.catalogFormId}?button=save`;
+
+    API.post(`/api/service_templates/${initialData.catalogFormId}`, { resource: requestObject, action: 'edit', skipErrors: [400] })
+      .then((_response) => {
+        const successMsg = sprintf(__('Catalog Item %s was saved'), values.name);
+        miqRedirectBack(successMsg, 'success', redirectUrl);
+      }).catch(({ data }) => {
+        const errorMessage = `${__(data.error.klass)}: ${__(data.error.message)}`;
+        miqRedirectBack(errorMessage, 'error', redirectUrl);
+      });
+  };
+
+  const mapper = {
+    ...componentMapper,
+    'key-value-list': KeyValueListComponent,
+    'copy-from-provisioning': CopyFromProvisonButton,
+    'tree-view-redux': TreeViewReduxWrapper,
+    'conditional-checkbox': conditionalCheckbox,
+  };
+
+  const customValidatorMapper = {
+    customValidatorForRetirementFields: () => (value, allValues) => {
+      if (!value && allValues.config_info.retirement.repository_id) {
+        return __('Required');
+      }
+      return false;
+    },
+
+    customValidatorForPrice: () => (value, allValues) => {
+      if (!value && allValues.currency_id) {
+        return __('Required');
+      }
+      return false;
+    },
+  };
+
+  return !data.isLoading && (
+    <MiqFormRenderer
+      onSubmit={onSubmit}
+      componentMapper={mapper}
+      validatorMapper={customValidatorMapper}
+      initialValues={data.formData}
+      onCancel={onCancel}
+      schema={createSchema({
+        availableCatalogs: data.availableCatalogs,
+        zones: data.zones,
+        currencies: data.currencies,
+        repositories: data.repositories,
+        provisionRepositoryId: data.provisionRepositoryId,
+        retirementRepositoryId: data.retirementRepositoryId,
+        setData,
+        machineCredentials: data.machineCredentials,
+        vaultCredentials: data.vaultCredentials,
+        cloudTypes: data.cloudTypes,
+        logOutputTypes: getLogOutputTypes(),
+        verbosityTypes: getVerbosityTypes(),
+        dialogs: data.dialogs,
+        provisionCloudType: data.provisionCloudType,
+        retirementCloudType: data.retirementCloudType,
+        tenantTree: initialData.tenantTree,
+        roleAllows: initialData.roleAllows,
+        provisionEsclationDisplay: data.provisionEsclationDisplay,
+        retirementEsclationDisplay: data.retirementEsclationDisplay,
+        currentRegion: initialData.currentRegion,
```

Comment:
Can we pass the `data` into the createSchema and pick the necessary items for the schema file. by doing so we can reduce the lines here.